external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaConfigUpdaterSuite.scala

```diff
@@ -0,0 +1,186 @@
+    )
+    val creds = new Credentials()
+    creds.addToken(KafkaTokenUtil.TOKEN_SERVICE, token)
+    UserGroupInformation.getCurrentUser.addCredentials(creds)
+  }
+
+  private def resetUGI(): Unit = {
+    UserGroupInformation.setLoginUser(null)
+  }
+
+  private def setSparkEnv(settings: Traversable[(String, String)]): Unit = {
+    val conf = new SparkConf().setAll(settings)
+    val env = mock(classOf[SparkEnv])
+    doReturn(conf).when(env).conf
+    SparkEnv.set(env)
+  }
+
+  private def resetSparkEnv(): Unit = {
+    SparkEnv.set(null)
+  }
+
+  test("set should always set value") {
+    val params = Map.empty[String, String]
+
+    val updatedParams = KafkaConfigUpdater(testModule, params)
+      .set(testKey, testValue)
+      .build()
+
+    assert(updatedParams.size() === 1)
+    assert(updatedParams.get(testKey) === testValue)
+  }
+
+  test("setIfUnset without existing key should set value") {
+    val params = Map.empty[String, String]
+
+    val updatedParams = KafkaConfigUpdater(testModule, params)
+      .setIfUnset(testKey, testValue)
+      .build()
+
+    assert(updatedParams.size() === 1)
+    assert(updatedParams.get(testKey) === testValue)
+  }
+
+  test("setIfUnset with existing key should not set value") {
+    val params = Map[String, String](testKey -> testValue)
+
+    val updatedParams = KafkaConfigUpdater(testModule, params)
+      .setIfUnset(testKey, otherTestValue)
+      .build()
+
+    assert(updatedParams.size() === 1)
+    assert(updatedParams.get(testKey) === testValue)
+  }
+
+  test("setAuthenticationConfigIfNeeded with global security should not set values") {
+    val params = Map.empty[String, String]
+    setGlobalKafkaClientConfig()
+
+    val updatedParams = KafkaConfigUpdater(testModule, params)
+      .setAuthenticationConfigIfNeeded()
+      .build()
+
+    assert(updatedParams.size() === 0)
+  }
+
+  test("setAuthenticationConfigIfNeeded with token should set values") {
+    val params = Map.empty[String, String]
+    setSparkEnv(Map.empty)
+    addTokenToUGI()
+
+    val updatedParams = KafkaConfigUpdater(testModule, params)
+      .setAuthenticationConfigIfNeeded()
+      .build()
+
+    assert(updatedParams.size() === 2)
+    assert(updatedParams.containsKey(SaslConfigs.SASL_JAAS_CONFIG))
+    assert(updatedParams.get(SaslConfigs.SASL_MECHANISM) === "SCRAM-SHA-512")
+  }
+
+  test("setAuthenticationConfigIfNeeded with token and invalid mechanism should throw exception") {
+    val params = Map.empty[String, String]
+    setSparkEnv(Map[String, String](Kafka.TOKEN_SASL_MECHANISM.key -> "INVALID"))
+    addTokenToUGI()
+
+    val e = intercept[IllegalArgumentException] {
+      KafkaConfigUpdater(testModule, params)
+        .setAuthenticationConfigIfNeeded()
+        .build()
+    }
+
+    assert(e.getMessage.contains("Delegation token works only with SCRAM mechanism."))
+  }
+
+  test("setAuthenticationConfigIfNeeded without security should not set values") {
```

Comment:
Just a question. So, `setAuthenticationConfigIfNeeded` is currently no-op in this case intentionally. Is it better to raise an `IllegalArgumentException` with some directional error message about requirements?

Reply:
This was the original implementation which was asked to change like this on other PRs (and I agree with it).
The reasoning behind was to keep fluent API and not having something like this(which is hardly testable):
```
    val updater = KafkaConfigUpdater("executor", specifiedKafkaParams)
      .set(...)
      .set(...)

    if (!globalSecurity && tokenAvailable)
      updater.setAuthenticationConfig(...)
    else if (somethingNew)
      updater.set(...)

    updater.build()
```
There is no flag like `kafka.securityEnabled` and Kafka can't tell it either.

> Is it better to raise an IllegalArgumentException with some directional error message about requirements?

Security parameters shouldn't always set, for example:
* If unsecure
* If user provided JVM global JAAS configuration (this provide total freedom to users)

> It would fail fast and gives us the chance to provide proper guide

* No security => nothing to check and do
* With global JAAS => it's not possible to parse JAAS config and suggest changes
* With DT => The user warned in advance to provide `SCRAM` related `sasl.mechanism` because the related Kafka error message is cryptic a bit.

I personally don't see more possibilities to provide convenience for users.

Reply:
I meant the case end users intend to set security so call `setAuthenticationConfig*` in KafkaConfigUpdater but they missed setting up security config. I guess error message will be provided in Kafka side but it might be cryptic so seeking the possibility to provide better message. 
No strong opinion and either is fine because end users can be indicated in any way. Just for clarifying my opinion.